File: cpp/epi/sort_students.cc
#include <algorithm>
#include <vector>
#include "fmt_print.h"
#include "generic_test.h"
#include "serialization_traits.h"
using std::string;
using std::vector;

struct Student {
  string name;
  float gpa;

  bool operator<(const Student& that) {
    return name < that.name;
  }
};

template <>
struct SerializationTraits<Student>
    : UserSerTraits<Student, string, float> {
  
  static std::vector<std::string> GetMetricNames(std::string& arg_name) {
    return {};
  }

  static std::vector<int> GetMetrics(const Student& student) {
    return {};
  }
};

bool operator==(const Student& lhs, const Student& rhs) {
  return (lhs.name == rhs.name) && (lhs.gpa == rhs.gpa);
}

std::ostream& operator<<(std::ostream& out, const Student& student) {
  return PrintTo(out, std::make_tuple(student.name, student.gpa));
}

vector<Student> SortStudents(const vector<Student>& students) {
  return {};
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::string func_name = "SortStudents";
  std::vector<std::string> param_names{"students"};

  return GenericTestMain(args, "sort_students.cc", "../test_data/epi/sort_students.tsv",
                         &SortStudents, DefaultComparator{}, func_name, param_names);
}

File: java/epi/SortStudents.java
package epi;
import test_framework.EpiTest;
import test_framework.EpiUserType;
import test_framework.GenericTest;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class SortStudents {

  @EpiUserType(ctorParams = {String.class, Float.class})
  public static class Student implements Comparable<Student> {
    public String name;
    public Float gpa;

    public Student(String name, Float gpa) {
      this.name = name;
      this.gpa = gpa;
    }

    // used by e.g. Collections.sort() with no custom comparator
    public int compareTo(Student that) {
      return this.name.compareTo(that.name);
    }

    @Override
    public boolean equals(Object that) {
      if (that == null || !(that instanceof Student)) {
        return false;
      }

      Student thatS = (Student) that;

      return this.name.compareTo(thatS.name) == 0 && this.gpa.compareTo(thatS.gpa) == 0;
    }

    @Override
    public String toString() {
      return "[" + "name: " + name + ", gpa: " + gpa + "]";
    }
  }

  @EpiTest(testDataFile = "../test_data/epi/sort_students.tsv")
  public static List<Student> sortStudentsByName(List<Student> students) {
    List<Student> ans = new ArrayList<>(students);

    Collections.sort(ans, new Comparator<Student>() {
      @Override
      public int compare(Student a, Student b) {
          return a.gpa.compareTo(b.gpa);
      }
    });

    return ans;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "SortStudents.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}
